backend/associations/tests/election/test_choice.py

```diff
@@ -0,0 +1,306 @@
+    def test_if_not_election_admin_then_error(self):
+        for user in ALL_USERS_EXCEPT_ELECTION_PDM:
+            self.login(user)
+            # 404 because the choice is not in the queryset.
+            self.assertStatusCode(
+                self.retrieve(20), 404, user_msg=f"Passed for {user}."
+            )
+            self.assertStatusCode(
+                self.create({"election": 20, "name": "Un choix"}),
+                403,
+                user_msg=f"Passed for {user}.",
+            )
+            self.assertStatusCode(
+                self.update(20, {"name": "Un choix"}),
+                404,
+                user_msg=f"Passed for {user}.",
+            )
+            self.assertStatusCode(self.destroy(20), 404, user_msg=f"Passed for {user}.")
+
+    # From now on, the tests will only deal with the election administrators.
+
+    ########
+    # LIST #
+    ########
+
+    def test_only_list_association_choices(self):
+        self.login("17election_biero")
+        res = self.list()
+        self.assertStatusCode(res, 200)
+        self.assertEqual(
+            res.data["count"],
+            Choice.objects.filter(election__association="biero").count(),
+        )
+        self.assertNotEqual(
+            res.data["results"],
+            Choice.objects.count(),
+            msg="This test is useless: add choices for another association than biero.",
+        )
+
+    ############
+    # RETRIEVE #
+    ############
+
+    def test_only_retrieve_association_choices(self):
+        self.login("17election_biero")
+
+        self.assertNotEqual(
+            Choice.objects.filter(election__association="biero").count(),
+            Choice.objects.count(),
+            msg="This test is useless: add choices for another association than biero.",
+        )
+
+        for choice in Choice.objects.all():
+            res = self.retrieve(choice.id)
+
+            if choice.election.association.id == "biero":
+                self.assertStatusCode(res, 200)
+            else:
+                self.assertStatusCode(res, 404)
+
+    ##########
+    # CREATE #
+    ##########
+
+    def test_can_create_for_future_election(self):
+        self.login("17election_pdm")
+        self.assertFalse(Election.objects.get(pk=20).started)
+
+        res = self.create({"election": 20, "name": "Un choix"})
+
+        self.assertEqual(res.status_code, 201, msg=res.data)
+        self.assertEqual(
+            1,
+            Choice.objects.filter(
+                election=20,
+                name="Un choix",
+                number_of_offline_votes=0,
+                number_of_online_votes=0,
+            ).count(),
+        )
+
+    def test_new_number_of_online_votes_is_zero(self):
+        self.login("17election_pdm")
+        self.assertFalse(Election.objects.get(pk=20).started)
+
+        res = self.create(
+            {
+                "election": 20,
+                "name": "Un choix",
+                "number_of_online_votes": 4,

```

Comment:
Shouldn't `number_of_online_votes` be read-only (but updated by the views) ?

Reply:
Yes, but that's the point of the test: `test_new_number_of_online_votes_is_zero`